File: cpp/leetcode/add_two_numbers.cc
#define _USE_MATH_DEFINES

#include <algorithm>
#include <cmath>
#include <cstdint>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode* next) : val(x), next(next) {}
};

class Solution {
   public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* current = new ListNode();
        ListNode* temp = current;
        int sum = 0, carry = 0, first, second;
        while (l1 != nullptr || l2 != nullptr || carry != 0) {
            if (l1 != nullptr) {
                first = l1->val;
                l1 = l1->next;
            } else {
                first = 0;
            }
            
            if (l2 != nullptr) {
                second = l2->val;
                l2 = l2->next;
            } else {
                second = 0;
            }
            sum = carry + first + second;
            current->next = new ListNode(sum % 10);
            carry = sum / 10;
            current = current->next;
        }

        return temp->next;
    }
};

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int t;
    cin >> t;
    while (t-- > 0) {
    }
    return 0;
}

File: rust/src/leetcode/add_two_numbers.rs
// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode {
            next: None,
            val,
        }
    }
}

struct Solution;

impl Solution {
    pub fn add_two_numbers(l1: Option<Box<ListNode>>, l2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut temp = Some(Box::new(ListNode::new(0)));
        let mut current = temp.as_mut();
        let mut sum;
        let mut carry = 0;

        let mut l1_current = l1;
        let mut l2_current = l2;

        while carry != 0 || l1_current.is_some() || l2_current.is_some() {
            sum = carry;
            if let Some(node) = l1_current {
                sum += node.val;
                l1_current = node.next;
            }
            if let Some(node) = l2_current {
                sum += node.val;
                l2_current = node.next;
            }
            carry = sum / 10;

            if let Some(node) = current {
                node.next = Some(Box::new(ListNode::new(sum % 10)));
                current = node.next.as_mut();
            }
        }
        temp.unwrap().next
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_add_two_numbers() {
        Solution::add_two_numbers(None, None);
    }
}
